### PR TITLE
MB-6901 Approve Move after TOO approves Shipment

### PR DESCRIPTION
**Pull request by monfresh**
## Description

When a TOO approves a shipment, there are default service items that get
created, and they are approved by default. When a move doesn't have any
service items in SUBMITTED status, the move's status should be
`Approved`, but in `mto_service_item_creator`, we were always setting
the move's status to `Approvals Requested`.

Instead, we should be iterating through all of the move's service items
to check their status, and only if at least one service item is in
`Submitted` status should we set the Move's status to
`Approvals Requested`. Otherwise, we should set it to `Approved`.

## Reviewer Notes

1. Sign in as a TOO
2. Click on a New Move in your queue
3. Approve the shipment and submit the move
4. Go back to your queue
5. Verify that the move's status is `Approved`

Is there anything you would like reviewers to give additional scrutiny?

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
make db_dev_e2e_populate
make server_run
make office_client_run
```

## Code Review Verification Steps

* [ ] If the change is risky, it has been tested in experimental before merging.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-6901) for this change

### CODE REVIEW

**Review by Ryan-Koch — APPROVED** [review]
Does the thing! Nice work on the fix!

File: pkg/services/mto_service_item/mto_service_item_creator.go [hunk 1]
```diff
@@ -262,16 +262,39 @@ func (o *mtoServiceItemCreator) CreateMTOServiceItem(serviceItem *models.MTOServ
...
-		// be updated to 'Approvals Requested' if it's not already in that state,
-		// which will let the TOO know they need to review it.
-		err = move.SetApprovalsRequested()
-		if err != nil {
-			return fmt.Errorf("%e", err)
+		// If there are no service items that are SUBMITTED then we need to change
+		// the move status to APPROVED
+		moveShouldBeApproved := true
+
+		for _, serviceItem := range move.MTOServiceItems {
```
Comment (duncan-truss) [inline pkg/services/mto_service_item/mto_service_item_creator.go:269]: This is using the move fetched way up at the top of the function before any service items are created.  That makes me think this has stale data if we've created service items in the meantime apart from the one passed in https://github.com/transcom/mymove/pull/6036/files#diff-3e70a15456b235362960a745c6b20d1576191f0c660790bf6fdedc193584d2b9L234

File: pkg/services/mto_service_item/mto_service_item_creator.go [hunk 2]
```diff
@@ -262,16 +262,39 @@ func (o *mtoServiceItemCreator) CreateMTOServiceItem(serviceItem *models.MTOServ
...
+				return fmt.Errorf("%#v %e", verrs, err)
+			}
 		}
-		verrs, err = txBuilder.UpdateOne(&move, nil)
-		if verrs != nil || err != nil {
-			return fmt.Errorf("%#v %e", verrs, err)
+		// If the move should not be APPROVED (due to one or more service items
+		// being in SUBMITTED status) then it needs to be set to APPROVALS REQUESTED
+		// so the TOO can review it.
+		if !moveShouldBeApproved {
```
Comment (duncan-truss) [inline pkg/services/mto_service_item/mto_service_item_creator.go:289]: Personal preference I would make this an else on the above if but the comment is important.  This is just a really complicated function to follow overall.

File: pkg/services/mto_service_item/mto_service_item_creator.go (hunk 1 above)
Comment (monfresh) [reply pkg/services/mto_service_item/mto_service_item_creator.go:269]: Yeah, good catch. I fetched it againt at first but the way I did it didn't work, and then I forgot to go back and try it again. I was thinking we could call a `Reload` function like ActiveRecord has, and it looks like Pop has a `Reload` function too.

File: pkg/services/mto_service_item/mto_service_item_creator.go (hunk 2 above)
Comment (monfresh) [reply pkg/services/mto_service_item/mto_service_item_creator.go:289]: Yeah, I'm working on refactoring it. I probably won't get to all of it by tomorrow, but I'll keep pushing incremental improvements.

File: pkg/services/mto_service_item/mto_service_item_creator.go (hunk 1 above)
Comment (monfresh) [reply pkg/services/mto_service_item/mto_service_item_creator.go:270]: Done.

File: pkg/services/mto_service_item/mto_service_item_creator.go [hunk 3]
```diff
@@ -262,16 +262,39 @@ func (o *mtoServiceItemCreator) CreateMTOServiceItem(serviceItem *models.MTOServ
...
-		// Once the service item is successfully created, the Move's status needs to
-		// be updated to 'Approvals Requested' if it's not already in that state,
-		// which will let the TOO know they need to review it.
-		err = move.SetApprovalsRequested()
-		if err != nil {
-			return fmt.Errorf("%e", err)
+		// If there are no service items that are SUBMITTED then we need to change
+		// the move status to APPROVED
+		moveShouldBeApproved := true
+		tx.Reload(&move)
```
Comment (duncan-truss) [inline pkg/services/mto_service_item/mto_service_item_creator.go:268]: Nice we could probably use this more if it's efficient.  I couldn't find it documented anywhere that it reloads associations but I tried it with a couple tests and I think it does do that properly.  Just wouldn't surprise me if that was another bug of theirs.

**Pull request merged**